lib/upipe-freetype/upipe_freetype.c

```diff
@@ -203,31 +239,57 @@ static struct upipe *upipe_freetype_alloc(struct upipe_mgr *mgr,
     if (unlikely(upipe == NULL))
         return NULL;
 
-    struct upipe_freetype *upipe_freetype = upipe_freetype_from_upipe(upipe);
+    upipe_freetype_init_urefcount(upipe);
+    upipe_freetype_init_output(upipe);
+    upipe_freetype_init_ubuf_mgr(upipe);
+    upipe_freetype_init_input(upipe);
 
+    struct upipe_freetype *upipe_freetype = upipe_freetype_from_upipe(upipe);
     upipe_freetype->flow_output = flow_def;
+    upipe_freetype->library = NULL;
+    upipe_freetype->cache_manager = NULL;
+    upipe_freetype->face = NULL;
+    upipe_freetype->xoff = 0;
+    upipe_freetype->yoff = 0;
+    upipe_freetype->font = NULL;
+    upipe_freetype->pixel_size = 0;
+
+    upipe_throw_ready(upipe);
 
     if (FT_Init_FreeType(&upipe_freetype->library)) {
-        uref_free(upipe_freetype->flow_output);
-        upipe_freetype_free_flow(upipe);
+        upipe_release(upipe);
         return NULL;
     }
 
-    upipe_freetype->face = NULL;
-
-    upipe_freetype_init_urefcount(upipe);
-    upipe_freetype_init_output(upipe);
-    upipe_freetype_init_ubuf_mgr(upipe);
-    upipe_freetype_init_input(upipe);
+    if (FTC_Manager_New(upipe_freetype->library,
+                        0, 0, 0,
+                        upipe_freetype_face_requester,
+                        upipe,
+                        &upipe_freetype->cache_manager)) {
+        upipe_release(upipe);
+        return NULL;
+    }
 
-    upipe_throw_ready(upipe);
+    if (FTC_CMapCache_New(upipe_freetype->cache_manager,
+                          &upipe_freetype->cmap_cache) ||
+        FTC_ImageCache_New(upipe_freetype->cache_manager,
+                           &upipe_freetype->img_cache) ||
+        FTC_SBitCache_New(upipe_freetype->cache_manager,
+                          &upipe_freetype->sbit_cache)) {
+        upipe_release(upipe);
+        return NULL;
+    }
```

Comment:
The Freetype context and caches should probably be shared among pipes, so shouldn't they be stored in the manager instead?

Reply:
Is it thread-safe ?

Reply:
No it's not, I'll have to add at least one mutex for the FT_Face and I'm not sure it's possible to use the same library with cache in multiple threads.